File: ReactF/ToolShapeGenerator.fs
namespace ReactF

module ToolShapeGenerator =
    open System
    open Serilog
    open ToolShape
    open System.Security.Cryptography

    let private rng = RandomNumberGenerator.Create()

    let private randomBytes count =
        try
            let bytes = Array.zeroCreate<byte> count
            rng.GetBytes(bytes)
            bytes
        with ex ->
            Log.Error(ex, "ToolShapeGenerator.randomBytes failed")
            Array.zeroCreate<byte> count

    let private randomCircle () =
        try
            let bytes = randomBytes 1
            bytes.[0] < 25uy
        with ex ->
            Log.Error(ex, "ToolShapeGenerator.randomCircle failed")
            false

    let private randomColor () =
        try
            let bytes = randomBytes 3
            sprintf "#%02X%02X%02X" bytes.[0] bytes.[1] bytes.[2]
        with ex ->
            Log.Error(ex, "ToolShapeGenerator.randomColor failed")
            "#000000"

    let private randomCoordinate max =
        try
            if max <= 0 then 0.0
            else
                let bytes = randomBytes 4
                let rand = BitConverter.ToUInt32(bytes, 0)
                (float rand / float UInt32.MaxValue) * float max
        with ex ->
            Log.Error(ex, "ToolShapeGenerator.randomCoordinate failed")
            0.0

    let private randomOpacity (min: float) (max: float) =
        try
            let bytes = randomBytes 4
            let rand = BitConverter.ToUInt32(bytes, 0)
            (float rand / float UInt32.MaxValue) * (max - min) + min
        with ex ->
            Log.Error(ex, "ToolShapeGenerator.randomOpacity failed")
            min

    let private randomRotation () =
        try
            let bytes = randomBytes 8
            let rand = BitConverter.ToUInt64(bytes, 0)
            (float rand / float UInt64.MaxValue) * 360.0
        with ex ->
            Log.Error(ex, "ToolShapeGenerator.randomRotation failed")
            0.0

    let private randomScale (min: float) (max: float) =
        try
            let bytes = randomBytes 4
            let rand = BitConverter.ToUInt32(bytes, 0)
            (float rand / float UInt32.MaxValue) * (max - min) + min
        with ex ->
            Log.Error(ex, "ToolShapeGenerator.randomScale failed")
            min

    let private randomSides (min: int) (max: int) =
        try
            if min > max then 12
            else
                let bytes = randomBytes 4
                let rand = BitConverter.ToUInt32(bytes, 0)
                int (rand % uint32 (max - min + 1)) + min
        with ex ->
            Log.Error(ex, "ToolShapeGenerator.randomSides failed")
            12

    let private randomStar () =
        try
            let bytes = randomBytes 1
            bytes.[0] < 50uy
        with ex ->
            Log.Error(ex, "ToolShapeGenerator.randomStar failed")
            false

    let generateShape (input: ToolShape option) : ToolShape option =
        try
            let i = input |> Option.bind (fun s -> s.id) |> Option.defaultValue (Guid.NewGuid())
            let x = input |> Option.bind (fun s -> s.x) |> Option.defaultValue (randomCoordinate 7648)
            let y = input |> Option.bind (fun s -> s.y) |> Option.defaultValue (randomCoordinate 4304)
            let circle = input |> Option.bind (fun s -> s.circle) |> Option.defaultValue (randomCircle ())
            let opacity = input |> Option.bind (fun s -> s.opacity) |> Option.defaultValue (randomOpacity 0.0 1.0)
            let rgb = input |> Option.bind (fun s -> s.rgb) |> Option.defaultValue (randomColor ())
            let rotation = input |> Option.bind (fun s -> s.rotation) |> Option.defaultValue (randomRotation ())
            let scale = input |> Option.bind (fun s -> s.scale) |> Option.defaultValue (randomScale 0.1 2.0)
            let sides = input |> Option.bind (fun s -> s.sides) |> Option.defaultValue (randomSides 3 8)
            let star = input |> Option.bind (fun s -> s.star) |> Option.defaultValue (randomStar ())
            Some {
                id = Some i
                x = Some x
                y = Some y
                circle = Some circle
                opacity = Some opacity
                rgb = Some rgb
                rotation = Some rotation
                scale = Some scale
                sides = Some sides
                star = Some star
            }
        with ex ->
            Log.Error(ex, "ToolShapeGenerator.generateShape failed")
            None

File: Reacts/Server/ToolShapeGenerator/ToolShapeGenerator.cs
namespace KusDepot.Reacts;

public static class ToolShapeGenerator
{
    private static readonly RandomNumberGenerator? _rng = RandomNumberGenerator.Create();

    public static ToolShape? GenerateShape(ToolShape? input = null)
    {
        try
        {
            return new ToolShape
            {
                id       = input?.id       ?? Guid.NewGuid(),
                x        = input?.x        ?? RandomCoordinate(7648),
                y        = input?.y        ?? RandomCoordinate(4304),
                circle   = input?.circle   ?? RandomCircle(),
                opacity  = input?.opacity  ?? RandomOpacity(),
                rgb      = input?.rgb      ?? RandomColor(),
                rotation = input?.rotation ?? RandomRotation(),
                scale    = input?.scale    ?? RandomScale(max:2.0),
                sides    = input?.sides    ?? RandomSides(3,8),
                star     = input?.star     ?? RandomStar()
            }.AsValid() ?? GenerateShape(input);
        }
        catch ( Exception _ ) { Log.Logger.Error(_,GenerateShapeFail); return null; }
    }

    private static Boolean RandomCircle()
    {
        try
        {
            var bytes = new Byte[1]; _rng!.GetBytes(bytes); return bytes[0] < 25;
        }
        catch ( Exception _ ) { Log.Logger.Error(_,RandomCircleFail); return false; }
    }

    private static String RandomColor()
    {
        try
        {
            var bytes = new Byte[3]; _rng!.GetBytes(bytes);

            return $"#{bytes[0]:X2}{bytes[1]:X2}{bytes[2]:X2}";
        }
        catch ( Exception _ ) { Log.Logger.Error(_,RandomColorFail); return String.Empty; }
    }

    private static Double RandomCoordinate(Int32 max)
    {
        try
        {
            if (max <= 0) { return 0; }

            var bytes = new Byte[4]; _rng!.GetBytes(bytes);

            UInt32 rand = BitConverter.ToUInt32(bytes,0);

            return (rand / (Double)UInt32.MaxValue) * max;
        }
        catch ( Exception _ ) { Log.Logger.Error(_,RandomCoordinateFail); return 0; }
    }

    private static Double RandomOpacity(Double min = 0.0 , Double max = 1.0)
    {
        try
        {
            var bytes = new Byte[4]; _rng!.GetBytes(bytes);

            UInt32 rand = BitConverter.ToUInt32(bytes,0);

            return (rand / (Double)UInt32.MaxValue) * (max - min) + min;
        }
        catch ( Exception _ ) { Log.Logger.Error(_,RandomOpacityFail); return 1; }
    }

    private static Double RandomRotation()
    {
        try
        {
            var bytes = new Byte[8]; _rng!.GetBytes(bytes);

            return (BitConverter.ToUInt64(bytes,0) / (Double)UInt64.MaxValue) * 360.0;
        }
        catch ( Exception _ ) { Log.Logger.Error(_,RandomRotationFail); return 0; }
    }

    private static Double RandomScale(Double min = 0.1 , Double max = 10.0)
    {
        try
        {
            var bytes = new Byte[4]; _rng!.GetBytes(bytes);

            UInt32 rand = BitConverter.ToUInt32(bytes,0);

            return (rand / (Double)UInt32.MaxValue) * (max - min) + min;
        }
        catch ( Exception _ ) { Log.Logger.Error(_,RandomScaleFail); return 1; }
    }

    private static Int32 RandomSides(Int32 min , Int32 max)
    {
        try
        {
            if(min > max) { return 12; }

            var bytes = new Byte[4]; _rng!.GetBytes(bytes);

            UInt32 rand = BitConverter.ToUInt32(bytes,0);

            return (Int32)(rand % (UInt32)(max - min + 1)) + min;
        }
        catch ( Exception _ ) { Log.Logger.Error(_,RandomSidesFail); return 12; }
    }

    private static Boolean RandomStar()
    {
        try
        {
            var bytes = new Byte[1]; _rng!.GetBytes(bytes); return bytes[0] < 50;
        }
        catch ( Exception _ ) { Log.Logger.Error(_,RandomStarFail); return false; }
    }
}